File: rust/src/play.rs
use world::World;
use std::time::Instant;

pub struct Play {}

impl Play {

  const World_Width: u32 = 150;
  const World_Height: u32 = 40;

  pub fn run() {
    let mut world = World::new(
      Play::World_Width,
      Play::World_Height
    );

    println!("{}", world.render());

    let mut total_tick: f64 = 0.0;
    let mut total_render: f64 = 0.0;

    loop {
      let tick_start = Instant::now();
      world._tick();
      let tick_time = tick_start.elapsed().as_secs_f64();
      total_tick += tick_time;
      let avg_tick = total_tick / world.tick as f64;

      let render_start = Instant::now();
      let rendered = world.render();
      let render_time = render_start.elapsed().as_secs_f64();
      total_render += render_time;
      let avg_render = total_render / world.tick as f64;

      let mut output = format!("#{}", world.tick);
      output += &format!(" - World tick took {} ({})", Play::_f(tick_time), Play::_f(avg_tick));
      output += &format!(" - Rendering took {} ({})", Play::_f(render_time), Play::_f(avg_render));
      output += &format!("\n{}", rendered.to_string());
      println!("{}", "\033[H\033[2J");
      println!("{}", output);
    }
  }

  pub fn _f(value: f64) -> String {
    return format!("{:.5}", value);
  }

}

File: zig/play.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const Io = std.Io;
const World = @import("world.zig").World;

pub const Play = struct {
    const WORLD_WIDTH = 150;
    const WORLD_HEIGHT = 40;

    pub fn run(allocator: Allocator, io: Io, minimal: bool) !void {
        const world = try World.init(allocator, io, WORLD_WIDTH, WORLD_HEIGHT);
        defer allocator.destroy(world);
        defer world.deinit();

        const stdout = Io.File.stdout();

        if (!minimal) {
            const rendered = try world.render();
            defer allocator.free(rendered);
            try stdout.writeStreamingAll(io, rendered);
        }

        var total_tick: f64 = 0;
        var lowest_tick = std.math.floatMax(f64);
        var total_render: f64 = 0;
        var lowest_render = std.math.floatMax(f64);

        while (true) {
            const tick_start = Io.Timestamp.now(io, .awake);
            world.dotick();
            const tick_finish = Io.Timestamp.now(io, .awake);
            const tick_time: f64 = @floatFromInt(tick_finish.nanoseconds - tick_start.nanoseconds);
            total_tick += tick_time;
            lowest_tick = @min(lowest_tick, tick_time);
            const avg_tick = total_tick / @as(f64, @floatFromInt(world.tick));

            const render_start = Io.Timestamp.now(io, .awake);
            const rendered = try world.render();
            defer allocator.free(rendered);
            const render_finish = Io.Timestamp.now(io, .awake);
            const render_time: f64 = @floatFromInt(render_finish.nanoseconds - render_start.nanoseconds);
            total_render += render_time;
            lowest_render = @min(lowest_render, render_time);
            const avg_render = total_render / @as(f64, @floatFromInt(world.tick));

            if (!minimal) {
                try stdout.writeStreamingAll(io, "\x1b[H\x1b[2J");
            }

            const output = try std.fmt.allocPrint(
                allocator,
                "#{d} - World Tick (L: {d:.3}; A: {d:.3}) - Rendering (L: {d:.3}; A: {d:.3})\n",
                .{
                    world.tick,
                    _f(lowest_tick),
                    _f(avg_tick),
                    _f(lowest_render),
                    _f(avg_render),
                },
            );
            defer allocator.free(output);
            try stdout.writeStreamingAll(io, output);

            if (!minimal) {
                try stdout.writeStreamingAll(io, rendered);
            }
        }
    }

    fn _f(value: f64) f64 {
        // nanoseconds -> milliseconds
        return value / 1_000_000.0;
    }
};

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const minimal = init.environ_map.contains("MINIMAL");

    try Play.run(allocator, io, minimal);
}
